src/core/lib/surface/call.c

```diff
@@ -1384,12 +1383,28 @@ static void receiving_initial_metadata_ready(grpc_exec_ctx *exec_ctx,
     }
   }
 
-  call->has_initial_md_been_received = true;
-  if (call->saved_receiving_stream_ready_bctlp != NULL) {
-    grpc_closure *saved_rsr_closure = GRPC_CLOSURE_CREATE(
-        receiving_stream_ready, call->saved_receiving_stream_ready_bctlp,
-        grpc_schedule_on_exec_ctx);
-    call->saved_receiving_stream_ready_bctlp = NULL;
+  grpc_closure *saved_rsr_closure = NULL;
+  while (true) {
+    gpr_atm rsr_bctlp =
+        gpr_atm_acq_load(&call->saved_receiving_stream_ready_bctlp);
+    /* Should only receive initial metadata once */
+    GPR_ASSERT(rsr_bctlp != 1);
+    if (rsr_bctlp == 0) {
+      /* Not received initial metadata and messages */
+      if (gpr_atm_no_barrier_cas(&call->saved_receiving_stream_ready_bctlp, 0,
```

Comment:
`no_barrier` should be enough for the previous implementation. As `receiving_initial_metadata_ready()` won't access the batch_control object from `receiving_stream_ready()` before and after the `cas` if the initial metadata is received first.
The `cas` in `receiving_stream_ready()`  is `rel`, as `receiving_initial_metadata_ready()` won't access the batch_control object before the `acq_load`.

In the new implementation, `receiving_initial_metadata_ready()` is using `acq_cas`, as it will only access the stored batch_control object after the `cas`. And its corresponding `cas` in ` receiving_stream_ready()` should be `rel`.

@sreecha I'm not very confident about the semantics used here. Could you please take a look?